src/Storages/ExternalStream/Kafka/KafkaSource.cpp

```diff
@@ -129,12 +136,27 @@ void KafkaSource::readAndProcess()
     current_batch.reserve(header.columns());
 
     Int64 current_batch_last_sn = -1;
-    auto callback = [&current_batch_last_sn, this](void * rkmessage, size_t total_count, void * data)
-    {
-        auto current_offset = parseMessage(rkmessage, total_count, data);
-        if (current_offset.has_value()) [[likely]]
-            current_batch_last_sn = *current_offset;
-    };
+    RdKafka::Consumer::Callback callback;
+    if (high_watermark == Kafka::NO_WARTERMARK)
```

Comment:
If we set NO_WATERMARK = std::numeric_limits<int64>() - 1; seems we can simply things even more.

We don't need branch to have 2 callbacks.

Reply:
Yes. However, that means it has to check high_wartermark against the current offset for each message, even in streaming mode. Having two callbacks is more code, but for streaming queries, it only check high_watermark for every message batch instead of every message.